Update ExternalMetrics provider to return error if datadogmetric is stale

sblumenthal opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

This PR will cause the external metrics server to return an error if a DatadogMetric object is valid but has not been updated within MaxAge.

### Motivation

See investigation [here](https://app.datadoghq.com/notebook/8582778/dca-serving-stale-datadogmetrics?notebookType=rte)

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

While this fixes the issue of the WPA scaling based off of metrics that have not been updated recently and have become stale, it does not actually mark the DatadogMetric CR object as invalid, so some confusion could present itself while looking at the cluster.

It also doesn't fix the underlying problem with the current implementation, where slow responses from the API lead to batch failures and batches are broken up and queries are evaluated sequentially, essentially halting all metric evaluation until the API recovers.

### Describe how to test/QA your changes

QA has already been done on my end via a minikube cluster where I spun up ~30 nginx deployments, each one configured to use a WPA for scaling. I modified the DCA using [this change](https://github.com/DataDog/datadog-agent/commit/8066cd80de8e643a396d99fb3d8356204ffd1c80) to simulate slow responses and timeouts, and recorded the results of the scaling behavior before and after the change in [this notebook](https://dddev.datadoghq.com/notebook/8833926/external-metrics-server-metric-staleness-testing)

sblumenthal committed:
Update ExternalMetrics provider to return error if datadogmetric is stale

sblumenthal committed:
Update docstring and error message

sblumenthal committed:
Update test expected error to match new format

sblumenthal committed:
Merge remote-tracking branch 'origin' into sblumenthal/CONTINT-4083

sblumenthal committed:
Add changelog

vboulineau commented on pkg/clusteragent/autoscaling/externalmetrics/model/datadogmetricinternal.go:
```diff
@@ -249,11 +259,15 @@ func (d *DatadogMetricInternal) BuildStatus(currentStatus *datadoghq.DatadogMetr
 		return nil, fmt.Errorf("DatadogMetric is invalid, err: %v", d.Error)
 	}
 
+	if d.IsStale(metricsMaxAge, time.Now().UTC().Unix()) {
```
currentTime should be injected as a parameter instead I believe

vboulineau reviewed (commented):
We should perhaps increase the default maxAge as we add the time from Controller (HPA,WPA) to ExternalMetrics, which can take several seconds.
Moving to 150s instead of 120s for instance.

sblumenthal committed:
Inject time of request to ToExternalMetricFormat, increase max_age default

sblumenthal commented:
> We should perhaps increase the default maxAge as we add the time from Controller (HPA,WPA) to ExternalMetrics, which can take several seconds. Moving to 150s instead of 120s for instance.

Synced about this offline, the reasoning behind this makes sense to me, so I updated the default value and added a comment about this change to the changelog

sblumenthal committed:
Revert max_age default back to 120s, add new config to account for delay

sblumenthal commented:
/merge

Pythyu commented:
/merge

Pythyu commented:
/merge

jennchenn commented:
/merge

jennchenn commented:
/merge
